### PR TITLE
New: Trakt Connection

### PR DESCRIPTION
**Pull request by geogolem**
This is the port of a similar feature recently added to Radarr.

https://github.com/geogolem/Radarr/commit/00f631c623e08699ed1c8dbb7e5566a1ab5c4152

#### Database Migration
NO

#### Description
Update Trakt Collection when item is added to Sonarr library

#### Todos
- [ ] Tests
- [ ] Documentation

#### Issues Fixed or Closed by this PR
https://github.com/Sonarr/Sonarr/issues/3906
*

### CODE REVIEW

**Comment by geogolem** [comment]
@rg9400 and @Qstick 
Both of you were involved with implementing this in Radarr and I used both your code/scripts to port it to Sonarr. Could you please take a look at the code and make sure it makes sense. It seems to be working. I don't think it effects anything in Sonarr, and see no reason why it couldnt eventually be merged into Sonarr. The only thing that needs to be done upon an actual merge of this PR is update the client ID to use a Sonarr clientID instead of the Radarr client ID.

**Comment by rg9400** [comment]
Payload construction looks correct, and if it is working for single and multi-episode files fine, then I think we are sending the info correctly.

The media object mappings all look good to me, including the switches for Sonarr-specific source. 

While mappings are not super important, there are two mappings worth highlighting:

- Source: Television -> DVD. I am personally fine mapping HDTV releases to DVD since there's no better option, but we can also map them to HD DVD for some level of differentiation (though that is not at all what the deprecated HD DVD format meant).
- Audio Channels: 0.0 -> Null. I know it's the same way in Radarr, but I think there's no realistic scenario where that is an accurate channel count. All such cases were related to issues in parsing Media Info for 1.0 PCM tracks.

File: src/NzbDrone.Core/Notifications/DownloadMessage.cs
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/DownloadMessage.cs:2]: Why?

File: src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs:28]: Use the one from the list PR

File: src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs:27]: Use the Sonarr Urls (in the list PR)

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (Qstick) [inline src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:248]: Sonarr

File: src/NzbDrone.Core/Notifications/DownloadMessage.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/DownloadMessage.cs:11]: Looks like this isn't used, nor is it valid since there could be more than one episode in a file. `EpisodeFile.Episodes` should be used instead.

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:61]: Nit: New line before return (applies to other cases below as well).

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:144]: I think we should put TV in the `vhs` media type, it looks to be the closest option and correctly separates it from DVD.

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:167]: You can use string interpolation instead of `string.Format`:

`traktResolution = $"hd_1080{scanIdentifier}";`

File: src/NzbDrone.Core/Notifications/DownloadMessage.cs
Comment (geogolem) [reply src/NzbDrone.Core/Notifications/DownloadMessage.cs:11]: yes, i did use episodeFile.Episodes in the end. I am guessing I just added this line in by accident and then didn't remove it - and it didn't give any build errors so I didn't notice. I'll remove it.

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (geogolem) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:61]: no problem to fix.

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (geogolem) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:144]: I agree, that it is good to differentiate DVD from other TV recordings, but VHS isn't entirely accurate either.

@rg9400 tried to contact Trakt Dev to inquire about adding more options. We aren't too sure if they will respond. IN the meantime I will change to VHS as per @markus101 's suggestion. It is easy enough to change in the future if better options arise.

Comment (geogolem) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:148]: https://support.trakt.tv/forums/188762-general/suggestions/41164723-more-collection-metadata-options

**Comment by geogolem** [comment]
There is another branch that builds upon this branch. https://github.com/geogolem/Sonarr/tree/onDeleteTraktConnection

This new branch implements the OnDelete to update the Trakt Collection when episodesFiles or Series are removed from Sonarr. It is also a port over from Radarr. A PR should be opened for it after this is merged. It is a bit more involved and involved a migration -> so it is probably best to keep it as a separate PR but it builds upon the stuff in this PR.

**Comment by rg9400** [comment]
FYI, Trakt added a few more audio codecs, so I refined the options in Radarr here https://github.com/Radarr/Radarr/pull/4960

Trakt is still discussing adding a format for TV which would be the last thing remaining (outside of MediaInfo AudioChannel nonsense) to get proper data flowing from the Arrs to Trakt.

File: src/NzbDrone.Core/Notifications/Trakt/TraktSettings.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/TraktSettings.cs:31]: These need to increment by 1, 0-4 for the 5 settings.

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (markus101) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:148]: Looks like it's been completed, but I don't know what exactly they added as their API docs don't seem to have been updated (the 'MEDIA OBJECT POST DATA' section).

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:254]: `MediaInfoFormatter.FormatAudioChannels` does this based on actual audio channel positions, we should use that here as well, if it's incorrect then we should enhance that there.

**Review by markus101 — CHANGES_REQUESTED** [review]
Looks like Trakt added more formats, but not clear if they added something for TV, also a couple more things to clean up.

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (rg9400) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:254]: Yes, Trakt added the below formats

`dolby_digital_plus_atmos` for EAC3 Atmos tracks
`mp2` for MP2 tracks
`ogg_opus` for OPUS tracks.

I already discussed the TV option with him, but he's still debating it as he feels Digital works. He said to give him some time to think about it.

There are some more complex bugs for MediaInfo AudioChannels. 

MediaInfo does not give channel info (just says "object based") for DTS-X tracks. See MediaInfo issue here https://github.com/MediaArea/MediaInfo/issues/299. These are parsed as 8.0 by the Arrs because of the lack of actual channel layout, but need to be sent as 7.1 to Trakt. Not sure if you'd want to bypass the parsing in the Formatter itself.

For 6.0 cases, it is due to the issue we were discussing in Radarr. You can see the preliminary discussion below, and I can give a detailed breakdown of my Radarr files showing what Radarr is parsing vs what MediaInfo is displaying. https://github.com/Radarr/Radarr/issues/4852

For 0,0 cases, I can give some examples as well. These seem to be PCM 1.0 tracks that are being parsed as 0.0 for an unknown reason. 0.0 is not a realistic channel layout (means the file is completely silent).

Comment (markus101) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:254]: > MediaInfo does not give channel info (just says "object based") for DTS-X tracks.

I don't see a reason to special case that here and not fix it where we format all audio tracks, then we display the correct thing in the UI as well.

The 6.0 cases are definitely special, but honestly I'm inclined to say those stay as-is until things are parsed correctly.

The 0.0 case would be a fallback after formatting.

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (geogolem) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:148]: It seems that Trakt folks are still debating whether or not they will go ahead with the TV option.
Thanks to @rg9400 for looking into this with the trakt Folks: https://forums.trakt.tv/t/more-metadata-options-for-collections/711/12

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (geogolem) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:258]: @markus101 Can we not just leave the temporary fix in until MediaInfo fixes it. Maybe just open a issue on Sonarr side that references the open Radarr issue https://github.com/Radarr/Radarr/issues/4852 and any future Radarr fix can then be ported over to Sonarr? I think there are already plans to get this fixed, either by other Radarr folks and/or mediaInfo fixes...

Please advise what I should do for this PR. 

Also regarding the TRAKT TV Option --> @rg9400 and I will keep on top of that and fix it should Trakt do anything about it.

Comment (markus101) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:258]: At the very least this logic should be ported to the formatter, it’s trying to do the same thing this is, but in a different spot.

Comment (geogolem) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:258]: I'll look into moving the logic soon.

File: src/NzbDrone.Core/Notifications/Trakt/Resource/TraktAuthRefreshResource.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/Resource/TraktAuthRefreshResource.cs:9]: Nit: Put a new line before the decorator, makes it easier to see things at a glance, looks like several files need the same treatment.

File: src/NzbDrone.Core/Notifications/Trakt/Resource/TraktEpisodeResource.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/Resource/TraktEpisodeResource.cs:8]: Same thing here.

File: src/NzbDrone.Core/Notifications/Trakt/Resource/TraktListResource.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/Resource/TraktListResource.cs:8]: And here.

File: src/NzbDrone.Core/Notifications/Trakt/Resource/TraktListResource.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/Resource/TraktListResource.cs:10]: Nit: I'd group up all the props that don't have decorators (`Type`, `Watchers` and `Revenue`)

File: src/NzbDrone.Core/Notifications/Trakt/Resource/TraktListResource.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/Resource/TraktListResource.cs:22]: 👍 on the new line here, unless you want to group it with the other props. Need to rename `Movie` to `Episode`.

**Review by markus101 — CHANGES_REQUESTED** [review]
Looking good, just a few minor formatting nits then I think we'll be good.

**Comment by samip5** [comment]
Ah, good thing that I checked here before starting to do it myself. 👍

File: src/NzbDrone.Core/Notifications/Trakt/Resource/TraktShowIdsResource.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/Resource/TraktShowIdsResource.cs:12]: This should get it's own file

File: src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs:44]: From debugging?

File: src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs:105]: Nit: Group up all the headers

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (markus101) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:258]: What's left to do here?

**Review by markus101 — CHANGES_REQUESTED** [review]
A few minor things left.

**Comment by rg9400** [comment]
The audio channel mappings can be removed whenever the latest MediaInfo changes from Radarr (https://github.com/Radarr/Radarr/pull/5352) are pulled over. That fixes all the various scenarios.

**Comment by bakerboy448** [comment]
I'd hold off and wait until https://github.com/Radarr/Radarr/issues/5404 is fixed; then pull the PR that fixes that as well

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (geogolem) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:258]: It looks like we are waiting for the following 2 items:
When Radarr/Radarr#5404 is fixed; we'll pull it's corresponding PR as well as Radarr/Radarr#5352

**Comment by geogolem** [comment]
> I'd hold off and wait until [Radarr/Radarr#5404](https://github.com/Radarr/Radarr/issues/5404) is fixed; then pull the PR that fixes that as well as [Radarr/Radarr#5352](https://github.com/Radarr/Radarr/pull/5352)

Sounds good - I will address Markus most recent comments and merge the items you mentioned when they are ready..

**Comment by markus101** [comment]
Looks like everything has been merged on the Radarr side.

**Comment by geogolem** [comment]
> Looks like everything has been merged on the Radarr side.

I have been a way for a while. I am going to spend the next few days.. catching up on all the work items I sort of abandoned or was waiting to be merged...

**Comment by markus101** [comment]
@geogolem any update?

**Comment by geogolem** [comment]
I have been away. but i will revisit this soon hopefully. i think its pretty much ready to merge. but i need to double check

**Comment by geogolem** [comment]
OK, I cleaned up /addressed the comments. Please let me know if you see anything that still needs to be done before merging.

Thanks

File: src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs:21]: Should be updated to remove the `api` subdomain: https://github.com/Sonarr/Sonarr/commit/6db24013592dddea5d9c0c4eff27b0da4071b646

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (markus101) [inline src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:163]: This doesn't need to be recreated each time and would be better as a HashSet, similar to how it's done for subtitle file extensions: https://github.com/Sonarr/Sonarr/blob/0a2b109a3fe101e260b623d0768240ef8b7a47ae/src/NzbDrone.Core/Extras/Subtitles/SubtitleFileExtensions.cs

**Review by markus101 — CHANGES_REQUESTED** [review]
One authentication issue and a small improvement to mapping resolution.

File: src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs
Comment (geogolem) [reply src/NzbDrone.Core/Notifications/Trakt/TraktProxy.cs:21]: will look into this in the next few days.. sort of busy right now, but will look into this and the other comments soon.. :)

File: src/NzbDrone.Core/Notifications/Trakt/TraktService.cs
Comment (geogolem) [reply src/NzbDrone.Core/Notifications/Trakt/TraktService.cs:163]: I'll work on fixing this up soon...

**Comment by geogolem** [comment]
I noticed that some onDelete notification code has been merged here: https://github.com/Sonarr/Sonarr/pull/4158

This completes some of the work from this branch: https://github.com/geogolem/Sonarr/commits/onDeleteTraktConnection

Please hold off on completing/merging this PR -- I am going to look into including the necessary things from that onDeleteTraktConnection so that there will be no need for an additional PR.. most of the changes from that branch have already been handled and are merged from the other PR - so its only the trakt stuff that needs to be added and it is probably best to just add it here and include in this PR...

**Comment by geogolem** [comment]
OK - so this should be ready for review and/or merged. Please let me know if there is anything else I can do to get this merged.

Thanks!

**Comment by JourneyOver** [comment]
Any word on when this is possibly coming to sonarr?

**Pull request merged**
